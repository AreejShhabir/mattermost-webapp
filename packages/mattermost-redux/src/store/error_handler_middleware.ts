// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {AnyAction, Dispatch, Middleware} from 'redux';

import {ThunkDispatch} from 'redux-thunk';

import {GlobalState} from '@mattermost/types/store';

import {UserTypes} from 'mattermost-redux/action_types';

import {Client4} from 'mattermost-redux/client';

import {isServerError} from '@mattermost/types/errors';
import {logError} from 'mattermost-redux/actions/errors';

type ErrorHandlerMiddleware = Middleware<

// This middleware doesn't add anything to dispatch, and this is the TS version of an empty object
Record<string, never>,

// Since forceLogoutIfNecessary relies on getState, we need to define the type of state needed by this middleware
GlobalState,

// Since we dispatch a non-object action (logError), TS needs to know that we're dispatching to redux-thunk
ThunkDispatch<GlobalState, undefined, AnyAction>
>;

export const errorHandlerMiddleware: ErrorHandlerMiddleware = ({dispatch, getState}) => (next) => (action) => {
    function handleError(error: unknown) {
        if (error instanceof Error) {
            forceLogoutIfNecessary(error, dispatch, getState);
            dispatch(logError(error));
        }

        return {error};
    }

    // Hand the action off to the next middleware (likely thunk) or dispatch it
    let result: any;
    try {
        result = next(action);
    } catch (error) {
        // If an error was thrown synchronously, handle and return it
        return handleError(error);
    }

    if (result instanceof Promise) {
        // This is an async action, so handle the result asynchronously
        return new Promise((resolve) => {
            // If the async call succeeds, return its result. If it fails, handle the error and return that.
            result.then(resolve).catch((error: unknown) => resolve(handleError(error)));
        });
    }

    return result;
};

const HTTP_UNAUTHORIZED = 401;

export function forceLogoutIfNecessary<D extends Dispatch<AnyAction> = Dispatch<AnyAction>>(err: Error, dispatch: D, getState: () => GlobalState) {
    const {currentUserId} = getState().entities.users;
    if (
        isServerError(err) &&
        err.status_code === HTTP_UNAUTHORIZED &&
        (err.url && err.url.indexOf('/login') === -1) &&
        currentUserId
    ) {
        Client4.setToken('');
        dispatch({type: UserTypes.LOGOUT_SUCCESS, data: {}});
    }
}

