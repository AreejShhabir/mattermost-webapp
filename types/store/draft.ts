// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {FileInfo} from '@mattermost/types/files';
import {PostType} from '@mattermost/types/posts';

export type PostDraft = {
    message: string;
    fileInfos: FileInfo[];
    uploadsInProgress: string[];
    props?: any;
    caretPosition?: number;
    postType?: PostType;
};
