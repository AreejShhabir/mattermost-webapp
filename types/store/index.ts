// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Channel} from '@mattermost/types/channels';
import {GlobalState as BaseGlobalState} from '@mattermost/types/store';
import {RelationOneToOne} from '@mattermost/types/utilities';

import {NewPostDraft} from './draft';
import {PluginsState} from './plugins';
import {ViewsState} from './views';

export type DraggingState = {
    state?: string;
    type?: string;
    id?: string;
}

export type GlobalState = BaseGlobalState & {
    plugins: PluginsState;
    storage: {
        storage: Record<string, any>;
        initialized: boolean;
    };
    views: ViewsState;
    drafts: {
        byChannel: RelationOneToOne<Channel, NewPostDraft>;
        byThread: RelationOneToOne<Channel, NewPostDraft>;
    };
};
