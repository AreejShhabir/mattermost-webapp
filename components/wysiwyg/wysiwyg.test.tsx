// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {NewPostDraft} from 'types/store/draft';

import {PostType} from '../file_upload';

import Wysiwyg, {WysiwygConfig} from './wysiwyg';

jest.mock('localforage', () => {
    const actual = jest.requireActual('localforage');

    return {
        ...actual,
        ready: jest.fn().mockResolvedValue(true),
        supports: jest.fn().mockResolvedValue(true),
    };
});

const wysiwygConfig: WysiwygConfig = {
    enterHandling: {
        ctrlSend: true,
        codeBlockOnCtrlEnter: true,
    },
    suggestions: {
        mention: {
            teamId: 'teamId',
            channelId: 'channelId',
            useSpecialMentions: false,
            useGroupMentions: false,
        },
        channel: {teamId: 'teamId'},
        command: {
            teamId: 'teamId',
            channelId: 'channelId',
        },
    },
    fileUpload: {
        rootId: '',
        channelId: 'channelId',
        postType: PostType.post,
    },
    enablePriority: false,
    enableEmojiPicker: false,
    useCustomEmojis: false,
    locale: 'en',
};

const props = {
    onSubmit: jest.fn(),
    onChange: jest.fn(),
    onAttachmentChange: jest.fn(),
    config: wysiwygConfig,
    readOnly: false,
    draft: {
        message: '',
    } as NewPostDraft,
    placeholder: 'Write to test-channel',
};

describe('WYSIWYG editor', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<Wysiwyg {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });
});
