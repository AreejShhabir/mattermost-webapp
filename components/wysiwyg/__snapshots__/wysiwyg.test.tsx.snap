// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WYSIWYG editor matches snapshot 1`] = `
<div
  configObservables={[Function]}
  getItemObservable={[Function]}
  newObservable={[Function]}
>
  NOTHING TO SEE HERE
</div>
`;
